compiler/src/iree/compiler/InputConversion/Common/QuantizedConvToConv.cpp

```diff
@@ -0,0 +1,320 @@
+  }
+
+  for (int i = 0, s = filterExprs.size(); i < s; i++) {
+    if (!filterExprs[i]) {
+      auto expr = mlir::getAffineDimExpr(iterators.size(), context);
+      iterators.push_back(getReductionIteratorTypeName());
+      filterExprs[i] = expr;
+    }
+  }
+
+  SmallVector<AffineMap> affineMaps{AffineMap::get(4, 0, filterExprs, context),
+                                    AffineMap::get(4, 0, outputExprs, context)};
+
+  return rewriter
+      .create<linalg::GenericOp>(
+          loc, zeroAcc.getType(), ValueRange{val}, ValueRange{zeroAcc},
+          affineMaps, iterators,
+          [=](OpBuilder &b, Location loc, ValueRange args) {
+            Value ext = b.create<arith::ExtSIOp>(loc, accETy, args[0]);
+            Value sum = b.create<arith::AddIOp>(loc, ext, args[1]);
+            b.create<linalg::YieldOp>(loc, sum);
+          })
+      .getResult(0);
+}
+
+// Pattern lowering conv_2d_nhwc_hwcf_q to conv_2d_nhwc_hwcf.
+//
+// This is implementing the math explained in Section 2.3 of
+// https://arxiv.org/abs/1712.05877.
+struct QuantizedConvToConv
+    : public OpRewritePattern<linalg::Conv2DNhwcHwcfQOp> {
+  using OpRewritePattern<linalg::Conv2DNhwcHwcfQOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfQOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    auto input = op.getInputs()[0];
+    auto filter = op.getInputs()[1];
+    auto iZp = op.getInputs()[2];
+    auto fZp = op.getInputs()[3];
+    auto inputTy = input.getType().cast<RankedTensorType>();
+    auto filterTy = filter.getType().cast<RankedTensorType>();
+    auto resultTy = op.getType(0).cast<ShapedType>();
+    auto accETy = resultTy.getElementType();
+
+    auto strides = op.getStrides();
+    auto dilations = op.getDilations();
+
+    IntegerAttr iZpConst;
+    IntegerAttr fZpConst;
+    bool iZpIsZero = matchPattern(iZp, m_Constant(&iZpConst)) &&
+                     iZpConst.getValue().isZero();
+    bool fZpIsZero = matchPattern(fZp, m_Constant(&fZpConst)) &&
+                     fZpConst.getValue().isZero();
+
+    // First implement the convolution without the zero point.
+    Value newConv = rewriter
+                        .create<linalg::Conv2DNhwcHwcfOp>(
+                            loc, resultTy, ValueRange{input, filter},
+                            op.getOutputs(), strides, dilations)
+                        .getResult(0);
+    RankedTensorType newConvTy = newConv.getType().cast<RankedTensorType>();
+
+    // If the zero pointer value is zero we can just replace.
+    if (iZpIsZero && fZpIsZero) {
+      rewriter.replaceOp(op, newConv);
+      return success();
+    }
+
+    // Compute the summation and correction for the filter zero point:
+    //  sum(d) = filter(a, b, c, d)
+    //  conv(a, b, c, d) -= iZp * filter(d)
+    if (!iZpIsZero) {
+      Value filterSum =
+          reduceDims(rewriter, loc, filter, accETy, {true, true, true, false});
```

Comment:
Could we do kwarg style comments for the true, true, true, false section? (a bit magical constant :))

Reply:
Added the kwarg descriptor.